File: ios/Classes/PdLonglinkPlugin.swift
import Flutter
import UIKit

public class PdLonglinkPlugin: NSObject, FlutterPlugin, FlutterStreamHandler, URLSessionWebSocketDelegate {
  private var eventSink: FlutterEventSink?
  private var nextSocketId: Int = 1
  private var tasks: [Int: URLSessionWebSocketTask] = [:]
  private var taskIdToSocketId: [Int: Int] = [:]

  private lazy var session: URLSession = {
    return URLSession(configuration: URLSessionConfiguration.default, delegate: self, delegateQueue: nil)
  }()

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "pd_longlink", binaryMessenger: registrar.messenger())
    let instance = PdLonglinkPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)

    let eventChannel = FlutterEventChannel(name: "pd_longlink/system_websocket_events", binaryMessenger: registrar.messenger())
    eventChannel.setStreamHandler(instance)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    case "systemWebSocket.connect":
      handleSystemWebSocketConnect(call, result: result)
    case "systemWebSocket.send":
      handleSystemWebSocketSend(call, result: result)
    case "systemWebSocket.sendBinary":
      handleSystemWebSocketSendBinary(call, result: result)
    case "systemWebSocket.close":
      handleSystemWebSocketClose(call, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }

  private func handleSystemWebSocketConnect(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "invalid_args", message: "arguments must be a map", details: nil))
      return
    }
    guard let urlString = args["url"] as? String, let url = URL(string: urlString) else {
      result(FlutterError(code: "invalid_args", message: "url is required", details: nil))
      return
    }

    let headers = args["headers"] as? [String: Any]
    let connectTimeoutMs = (args["connectTimeoutMs"] as? NSNumber)?.doubleValue ?? 20000

    var request = URLRequest(url: url)
    request.timeoutInterval = connectTimeoutMs / 1000.0
    headers?.forEach { (key: String, value: Any) in
      request.setValue("\(value)", forHTTPHeaderField: key)
    }

    let socketId = nextSocketId
    nextSocketId += 1

    let task = session.webSocketTask(with: request)
    tasks[socketId] = task
    taskIdToSocketId[task.taskIdentifier] = socketId
    task.resume()
    receiveLoop(socketId: socketId, task: task)
    result(["socketId": socketId])
  }

  private func handleSystemWebSocketSend(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "invalid_args", message: "arguments must be a map", details: nil))
      return
    }
    guard let socketId = args["socketId"] as? Int, let text = args["text"] as? String else {
      result(FlutterError(code: "invalid_args", message: "socketId and text are required", details: nil))
      return
    }
    guard let task = tasks[socketId] else {
      result(nil)
      return
    }
    task.send(.string(text)) { error in
      if let error = error {
        self.sendEvent(["socketId": socketId, "type": "error", "error": error.localizedDescription])
      }
      result(nil)
    }
  }

  private func handleSystemWebSocketSendBinary(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "invalid_args", message: "arguments must be a map", details: nil))
      return
    }
    guard let socketId = args["socketId"] as? Int, let bytes = args["bytes"] as? FlutterStandardTypedData else {
      result(FlutterError(code: "invalid_args", message: "socketId and bytes are required", details: nil))
      return
    }
    guard let task = tasks[socketId] else {
      result(nil)
      return
    }
    task.send(.data(bytes.data)) { error in
      if let error = error {
        self.sendEvent(["socketId": socketId, "type": "error", "error": error.localizedDescription])
      }
      result(nil)
    }
  }

  private func handleSystemWebSocketClose(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "invalid_args", message: "arguments must be a map", details: nil))
      return
    }
    guard let socketId = args["socketId"] as? Int else {
      result(FlutterError(code: "invalid_args", message: "socketId is required", details: nil))
      return
    }
    let code = (args["code"] as? Int) ?? 1000
    let reason = (args["reason"] as? String) ?? ""

    if let task = tasks.removeValue(forKey: socketId) {
      taskIdToSocketId.removeValue(forKey: task.taskIdentifier)
      task.cancel(with: URLSessionWebSocketTask.CloseCode(rawValue: code) ?? .normalClosure, reason: reason.data(using: .utf8))
    }
    result(nil)
  }

  private func receiveLoop(socketId: Int, task: URLSessionWebSocketTask) {
    task.receive { res in
      switch res {
      case .failure(let error):
        self.tasks.removeValue(forKey: socketId)
        self.sendEvent(["socketId": socketId, "type": "error", "error": error.localizedDescription])
      case .success(let message):
        switch message {
        case .string(let text):
          self.sendEvent(["socketId": socketId, "type": "message", "data": text])
        case .data(let data):
          self.sendEvent([
            "socketId": socketId,
            "type": "message",
            "isBinary": true,
            "dataBytes": FlutterStandardTypedData(bytes: data)
          ])
        @unknown default:
          break
        }
        self.receiveLoop(socketId: socketId, task: task)
      }
    }
  }

  public func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didOpenWithProtocol protocol: String?) {
    if let socketId = taskIdToSocketId[webSocketTask.taskIdentifier] {
      sendEvent(["socketId": socketId, "type": "open"])
    }
  }

  public func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didCloseWith closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?) {
    if let socketId = taskIdToSocketId[webSocketTask.taskIdentifier] {
      tasks.removeValue(forKey: socketId)
      taskIdToSocketId.removeValue(forKey: webSocketTask.taskIdentifier)
      let reasonText = reason.flatMap { String(data: $0, encoding: .utf8) } ?? ""
      sendEvent(["socketId": socketId, "type": "closed", "code": closeCode.rawValue, "reason": reasonText])
    }
  }

  private func sendEvent(_ event: [String: Any]) {
    eventSink?(event)
  }
}

File: android/src/main/kotlin/com/pedro/pd_longlink/PdLonglinkPlugin.kt
package com.pedro.pd_longlink

import android.os.Handler
import android.os.Looper
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.EventChannel
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener

class PdLonglinkPlugin :
    FlutterPlugin,
    MethodCallHandler,
    EventChannel.StreamHandler {

    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    private val nextSocketId: AtomicInteger = AtomicInteger(1)
    private val sockets: ConcurrentHashMap<Int, WebSocket> = ConcurrentHashMap()
    private val defaultClient: OkHttpClient = OkHttpClient.Builder().readTimeout(0, TimeUnit.MILLISECONDS).build()

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "pd_longlink")
        channel.setMethodCallHandler(this)
        eventChannel =
            EventChannel(flutterPluginBinding.binaryMessenger, "pd_longlink/system_websocket_events")
        eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result
    ) {
        when (call.method) {
            "getPlatformVersion" -> result.success("Android ${android.os.Build.VERSION.RELEASE}")
            "systemWebSocket.connect" -> handleSystemWebSocketConnect(call, result)
            "systemWebSocket.send" -> handleSystemWebSocketSend(call, result)
            "systemWebSocket.sendBinary" -> handleSystemWebSocketSendBinary(call, result)
            "systemWebSocket.close" -> handleSystemWebSocketClose(call, result)
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        eventSink = null
        sockets.values.forEach { ws -> ws.cancel() }
        sockets.clear()
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    private fun handleSystemWebSocketConnect(call: MethodCall, result: Result) {
        val args = call.arguments as? Map<*, *>
        val url = args?.get("url") as? String
        if (url.isNullOrEmpty()) {
            result.error("invalid_args", "url is required", null)
            return
        }

        val headers = args["headers"] as? Map<*, *>
        val connectTimeoutMs = (args["connectTimeoutMs"] as? Number)?.toLong() ?: 20000L

        val socketId = nextSocketId.getAndIncrement()
        val reqBuilder = Request.Builder().url(url)
        headers?.forEach { entry ->
            val k = entry.key?.toString() ?: return@forEach
            val v = entry.value?.toString() ?: return@forEach
            reqBuilder.addHeader(k, v)
        }

        val client =
            defaultClient.newBuilder().connectTimeout(connectTimeoutMs, TimeUnit.MILLISECONDS).build()

        val ws = client.newWebSocket(reqBuilder.build(), object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                sendEvent(mapOf("socketId" to socketId, "type" to "open"))
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                sendEvent(mapOf("socketId" to socketId, "type" to "message", "data" to text))
            }

            override fun onMessage(webSocket: WebSocket, bytes: ByteArray) {
                sendEvent(mapOf(
                    "socketId" to socketId,
                    "type" to "message",
                    "isBinary" to true,
                    "dataBytes" to bytes.toList()
                ))
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                sockets.remove(socketId)
                sendEvent(mapOf("socketId" to socketId, "type" to "closed", "code" to code, "reason" to reason))
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                sockets.remove(socketId)
                sendEvent(mapOf("socketId" to socketId, "type" to "error", "error" to t.toString()))
            }
        })

        sockets[socketId] = ws
        result.success(mapOf("socketId" to socketId))
    }

    private fun handleSystemWebSocketSend(call: MethodCall, result: Result) {
        val args = call.arguments as? Map<*, *>
        val socketId = (args?.get("socketId") as? Number)?.toInt()
        val text = args?.get("text") as? String
        if (socketId == null || text == null) {
            result.error("invalid_args", "socketId and text are required", null)
            return
        }

        sockets[socketId]?.send(text)
        result.success(null)
    }

    private fun handleSystemWebSocketSendBinary(call: MethodCall, result: Result) {
        val args = call.arguments as? Map<*, *>
        val socketId = (args?.get("socketId") as? Number)?.toInt()
        val rawBytes = args?.get("bytes")
        if (socketId == null || rawBytes == null) {
            result.error("invalid_args", "socketId and bytes are required", null)
            return
        }
        val bytes: ByteArray = when (rawBytes) {
            is ByteArray -> rawBytes
            is List<*> -> rawBytes.map { (it as? Number)?.toByte() ?: 0 }.toByteArray()
            else -> {
                result.error("invalid_args", "bytes must be ByteArray or List", null)
                return
            }
        }

        sockets[socketId]?.send(bytes)
        result.success(null)
    }

    private fun handleSystemWebSocketClose(call: MethodCall, result: Result) {
        val args = call.arguments as? Map<*, *>
        val socketId = (args?.get("socketId") as? Number)?.toInt()
        if (socketId == null) {
            result.error("invalid_args", "socketId is required", null)
            return
        }

        val code = (args["code"] as? Number)?.toInt() ?: 1000
        val reason = args["reason"] as? String ?: ""
        val ws = sockets.remove(socketId)
        if (ws != null) {
            ws.close(code, reason)
        }
        result.success(null)
    }

    private fun sendEvent(event: Map<String, Any?>) {
        mainHandler.post {
            eventSink?.success(event)
        }
    }
}